### DISCUSSION TITLE
Container-based solution matching security requirements.

### DISCUSSION THREAD

**Discussion by cerebrosius** [zvonkok/kata-containers#3] [Q&A] [answered]
How can we secure containers?

**Comment by zvonkok** [comment, marked as answer]
Traditional containers (Docker, LXC, rkt) are not truly sandboxed as they share the host OS kernel. Such containers are a unique way of packaging software to provide user-space independence. The problem is that a container breakout can compromise the whole node, even a complete Kubernetes cluster. The isolation techniques for traditional containers are cgroups, namespaces, and Kernel-related security features like SELinux, AppArmor,  and seccomp.


In the last few years, there have been many approaches to fortify the isolation (weak security boundary) between containers and the host OS to prevent container escapes motivated by a long list of CVEs CVE-2014-3519, CVE-2016-5195, CVE-2016-9962, CVE-2017-5123, CVE-2019-5736, CVE-2022-0492, CVE-2022-0185, CVE-2022-30137, that can lead to container breakouts. 

 

Most of the time, stronger trust boundaries for containers involve creating a hybrid architecture leveraging the strong trust boundaries of hardware virtualization, aka VMs, and focusing on better container efficiency. Some solutions are virtualizing parts of the OS to implement trust boundaries without VMs (usually, syscall virtualization).
